sklearn/preprocessing/imputation.py

```diff
@@ -413,6 +431,21 @@ def transform(self, X):
         valid_idx = np.where(valid_mask)[0]
         missing = np.arange(X.shape[not self.axis])[invalid_mask]
 
+        if sparse.issparse(X) and self.missing_values != 0:
+            # sparse matrix and missing values is not zero
+            mask = _get_mask(X.data, self.missing_values)
+            indexes = np.repeat(np.arange(len(X.indptr) - 1, dtype=np.int),
+                                np.diff(X.indptr))[mask]
+            self.mask_matrix_ = X.__class__((mask, X.indices.copy(),
```

Comment:
I remember that we should not set any attributes during transform time, then would it be okay to pass the ``mask_matrix_/mask_`` as an argument for ``sparse_transform_/dense_transform_`` from ``transform`` ?